Functionality to run ScanConsistencyIT standalone

kevinrr888 opened the pull request:
Closes #3646.
Refactored ScanConsistencyIT to be able to be run standalone. Can now be run as a test or from main.

keith-turner commented on test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java:
```diff
@@ -104,69 +150,97 @@ public void testConcurrentScanConsistency() throws Exception {
+    checkEquals(stats3.scanned, stats1.scanned);
+    checkEquals(stats3.verified, stats1.verified);
+
+    executor.shutdownNow();
```
Would be good to shutdown this executor in a finally block still.

keith-turner commented on test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java:
```diff
@@ -104,69 +150,97 @@ public void testConcurrentScanConsistency() throws Exception {
+      var stats = writeTask.get();
+      log.debug(String.format("Wrote:%,d Bulk imported:%,d Deleted:%,d Bulk deleted:%,d",
+          stats.written, stats.bulkImported, stats.deleted, stats.bulkDeleted));
+      checkTrue(stats.written + stats.bulkImported > 0);
```
Do you know what happens if assertTrue is called outside of a junit test?

kevinrr888 replied:
The program fails with a ClassNotFoundException for the class org.junit.jupiter.api.Assertions

DomGarguilo commented on test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java:
```diff
@@ -79,12 +79,58 @@
+     * "accumulo org.apache.accumulo.test.ScanConsistencyIT <props-file> <tmp-dir> <table>"
+     *      <props-file>: An accumulo client properties file
+     *      <tmp-dir>: tmpDir field for the TestContext object
+     *      <table>: The name of the table to be created
```
When running standalone, do we want a new table to be created? Or do we want to be able to pass in an existing table?

keith-turner replied:
The test would need an empty table if it existed.  Would probably be good to create the table and have the test fail if it exists.  The call to create table should fail if it exists, so should not be anything to do there.  Could delete the table when the test is done.

DomGarguilo commented on test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java:
```diff
@@ -104,69 +150,97 @@ public void testConcurrentScanConsistency() throws Exception {
+   *
+   * @param l1 One of the two longs to be compared
+   * @param l2 One of the two longs to be compared
+   * @throws Exception
```
I think these `@throws` tags without a description are causing the build to fail. I think they can be removed 
```suggestion
   * @throws Exception
```

DomGarguilo commented on test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java:
```diff
@@ -79,12 +79,58 @@
+      final String tmpDir = args[1];
+      final String table = args[2];
+
+      try {
+        AccumuloClient client = Accumulo.newClient().from(propsFile).build();
```
```suggestion
      try {
        AccumuloClient client = Accumulo.newClient().from(propsFile).build();
```
Can move `client` into the try-with-resources block so it is closed automatically.

DomGarguilo commented on test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java:
```diff
@@ -104,69 +150,97 @@ public void testConcurrentScanConsistency() throws Exception {
+    var stats1 = scanData(testContext, random, new Range(), false);
+    var stats2 = scanData(testContext, random, new Range(), true);
+    var stats3 = batchScanData(testContext, new Range());
+    log.debug(
```
When running this test standalone, it would be nice to see these logs at the `info` level instead of `debug`. Its not much of a hassle to move my instance to display the debug logs, bug when I do they get drowned out by a bunch of other debug logs.

I know these logs were not added in this PR but they could be changed since its related to the new context that this code is run.

keith-turner replied:
> bug when I do they get drowned out by a bunch of other debug logs.

Should be able to enable debug logging just for this class in log4j2 config.

Changing the log level to info in the code should be ok too.

kevinrr888 commented:
* Removed "@ throws Exception" from javadoc. Hopefully fixes build failure from "Checkstyle violations".
* Shutdown the executor in a finally block
* Table is now deleted when test is done
* Check args length using Preconditions.checkArgument()

keith-turner commented on test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java:
```diff
@@ -150,75 +147,77 @@ private static void runTest(AccumuloClient client, FileSystem fileSystem, String
+      checkEquals(stats3.verified, stats1.verified);
+    } finally {
+      executor.shutdownNow();
+      client.tableOperations().delete(table);
```
Placing code in a finally block that may throw an exception is tricky.  I would not expected shutting down the executor throw an exception, but deleting the table might.  If code in the try block throws an exception and code in the finally block throws an exception, then the exception from the try block is lost.  Not sure of a good way to improve this though other than something like the following.

```
try{

} catch(Exception e) {
   // log here in case the finally block also throws an exception
    log.error(e.getMessage(),e);
    throw e;
} finally {
   executor.shutdownNow();
   client.tableOperations().delete(table);
}

kevinrr888 replied:
Ah I see. Could I also just move the deletion of the table into the above try block?

kevinrr888 commented on test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java:
```diff
@@ -242,6 +241,19 @@ private static void checkEquals(long l1, long l2) throws Exception {
+   *
+   * @param msg message to be logged
+   */
+  private static void logMessage(String msg) {
```
Assumed (based on previous comments) that the message should be logged in debug mode if we are in the testing context, and info mode otherwise. Should this always be log.info?

EdColeman replied:
My reading is that is one possibility.  Depending on the frequency of some of the messages, maybe some are debug and others are at info?

Then using log4j configuration can be tuned for the circumstance.

keith-turner replied:
Changing everything to info should be fine.  The test are already configured to log debug output [here](https://github.com/apache/accumulo/blob/main/test/src/main/resources/log4j2-test.properties), so changing to info would not increase the amount output by the test when running it from maven.

dlmarion commented:
The same can be achieved using the instructions [here](https://github.com/apache/accumulo/blob/main/TESTING.md#integration-tests). Just curious, why would need this?

keith-turner commented:
> The same can be achieved using the instructions [here](https://github.com/apache/accumulo/blob/main/TESTING.md#integration-tests). Just curious, why would we need this?

We may not, I had forgotten about that, wonder if the following work

 1. Setup accumulo in uno
 2. copy test jar to uno lib dir, this is needed because the test has server side code it runs
 3. restart accumulo
 4. run test using mvn passing in uno accumulo props file

If it does work may still want some of the changes in this PR like deleting the table.

keith-turner commented:
When running this test manually against an existing instance, would like to be able to adjust [this sleep](https://github.com/apache/accumulo/blob/cb1a9ac9373b0cae311e5a67021666e14b006430/test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java#L132).  Would like to be able to run this test against an uno instance for hours.

dlmarion commented:
See [this](https://github.com/apache/accumulo/blob/main/TESTING.md#standalone-cluster-standalonecapablecluster) for running an IT against an existing instance. Caveat: I've never tried it.

kevinrr888 commented:
* Table deletion is no longer in finally block
* Use log.info
* Main now accepts sleep time

EdColeman commented on test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java:
```diff
@@ -170,34 +173,34 @@ private static void runTest(AccumuloClient client, FileSystem fileSystem, String
       for (Future<WriteStats> writeTask : writeTasks) {
         var stats = writeTask.get();
-        logMessage(String.format("Wrote:%,d Bulk imported:%,d Deleted:%,d Bulk deleted:%,d",
+        log.info(String.format("Wrote:%,d Bulk imported:%,d Deleted:%,d Bulk deleted:%,d",
```
These could use the logging placeholder `{ }` instead of String.format. It does not look like there is any specific conversions or padding.

kevinrr888 replied:
`%,d` (note the comma) converts the integer to be formatted with commas. I don't think `{}` can be used to format the string like this but I'm not sure

keith-turner commented on test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java:
```diff
@@ -79,12 +79,57 @@
+     * @formatter:off
+     * Note: In order to run main,
+     * 1) Build the project
+     * 2) Copy the accumulo test jar (in /test/target/) to your accumulo installation's
```
I think we could further minimize changes to the test by requiring copying any needed dependencies like junit

```
mvn dependency:copy-dependencies -DincludeGroupIds="org.junit.jupiter"
cp test/target/dependency/junit-jupiter-* $ACCUMULO_HOME/lib/
```

If that works, then would not need the wrappers around the junit assert method call, which would minimizes the test changes.

ctubbsii commented on test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java:
```diff
@@ -80,11 +80,58 @@ public class ScanConsistencyIT extends AccumuloClusterHarness {
-      justification = "predictable random is ok for testing")
+  public static void main(String[] args) {
+    /**
+     * @formatter:off
```
This is a javadoc. It doesn't make sense to turn off formatting for javadocs... because the end result is still rendered HTML, not formatted plain text. Instead, use html tags, like `<p>`, `<ul>`, etc. If you do it right, the javadoc will be formatted reasonably so that it's readable.

Also, if you want `<` and `>` to render properly, like you have near the bottom of this javadoc, you're going to need to express it as `&lt;` or `&gt;`.

kevinrr888 replied:
Thanks, I updated the comment

ctubbsii commented on test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java:
```diff
@@ -80,11 +80,58 @@ public class ScanConsistencyIT extends AccumuloClusterHarness {
+    try (AccumuloClient client = Accumulo.newClient().from(propsFile).build()) {
+      FileSystem fileSystem = FileSystem.get(new Configuration());
+      runTest(client, fileSystem, tmpDir, table, sleepTime);
+    } catch (Exception e) {
+      log.error(e.toString());
+    }
```
No need to catch and log. Just let it get thrown through the main method. The stack trace there will be more useful than the truncated exception that goes to the logger.

kevinrr888 replied:
Thanks, changed so main just throws an error now

ctubbsii commented on test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java:
```diff
@@ -80,30 +80,34 @@ public class ScanConsistencyIT extends AccumuloClusterHarness {
+   * [tmp-dir]: tmpDir field for the TestContext object<br>
+   * [table]: The name of the table to be created<br>
+   * [sleep-time]: The time to sleep (ms) after submitting the various concurrent tasks<br>
+   * <br>
```
I don't know if this works, but I was thinking something more along this:

```suggestion
   * Note: In order to run main,
   * <ol>
   * <li>Build the project
   * <li>Copy the accumulo test jar (in /test/target/) into your accumulo installation's lib directory
   * <li>Copy the JUnit dependencies into your accumulo installation's lib directory:
   * <ul>
   * <li>mvn dependency:copy-dependencies -DincludeGroupIds="org.junit.jupiter"
   * <li>cp test/target/dependency/junit-jupiter-* $ACCUMULO_HOME/lib/
   * </ul>
   * <li>Ensure the test jar is in lib before the tablet servers start. Restart tablet servers if
   * necessary.
   * <li>Run with: bin/accumulo org.apache.accumulo.test.ScanConsistencyIT [props-file] [tmp-dir] [table]
   * [sleep-time]
   * </ol>
   *
   * [props-file]: An accumulo client properties file<br>
   * [tmp-dir]: tmpDir field for the TestContext object<br>
   * [table]: The name of the table to be created<br>
   * [sleep-time]: The time to sleep (ms) after submitting the various concurrent tasks
```

kevinrr888 replied:
Reformatted similarly. From running the QA checks, maven didn't like the unordered list nested within the ordered list which is why I have the third list element the way it is.
